runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -59,16 +56,188 @@ extern "C" {
+        return NULL;
+    }
+    metaData->thunkAddress = (void *)thunkMem;
+
+    I_32 gprIndex = 1;
+    const I_32 maxGPRIndex = 3;
+    I_32 fprIndex = 0;
+    const I_32 maxFPRIndex = 6;
+    I_32 currArgIndex = 0;
+
+    // Now we loop through the arguments and generate the actual instructions to store any register arguments in the caller frame.
+    const I_32 numRegsToPreserve = 10;
+    const I_32 offsetToRegSaveArea = defaultStackPointerBias - thunkFrameSize + registerSaveAreaSize - (numRegsToPreserve * 8); // This should evaluate to 1808.
+    thunkMem += STMG(thunkMem, R6, R15, R4, offsetToRegSaveArea);
+    thunkMem += LAY(thunkMem, R4, 0, R4, -thunkFrameSize);
+    for (I_32 i = 0; i < lastSigIdx; i++) {
+        switch(sigArray[i].type) {
+            case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+            case J9_FFI_UPCALL_SIG_TYPE_INT64:
+                if (gprIndex <= maxGPRIndex) {
+                    thunkMem += STG(thunkMem, gprIndex, 0, R4, argAreaOffsetFromSP + (currArgIndex * 8));
+                }
+                currArgIndex += 1;
+                gprIndex += 1;
+                break;
+            case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+                if (fprIndex <= maxFPRIndex) {
+                    thunkMem += STEY(thunkMem, fprIndex, 0, R4, argAreaOffsetFromSP + (currArgIndex * 8));
+                }
+                currArgIndex += 1;
+                fprIndex += 2;
+                if (i < 3) {
+                    gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+                }
+                break;
+            case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+                if (fprIndex <= maxFPRIndex) {
+                    thunkMem += STDY(thunkMem, fprIndex, 0, R4, argAreaOffsetFromSP + (currArgIndex * 8));
+                }
+                currArgIndex += 1;
+                fprIndex += 2;
+                if (i < 3) {
+                    gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+                }
+                break;
+            default:
+                Assert_VM_unreachable();
+        }
+
+    }
+
+    // Now jump to the common dispatcher helper routine.
+    thunkMem += IIHF(thunkMem, R1, ((reinterpret_cast<uint64_t>(metaData)) >> 32));
+    thunkMem += IILF(thunkMem, R1, reinterpret_cast<uint64_t>(metaData));
+    thunkMem += LAY(thunkMem, R2, 0, R4, argAreaOffsetFromSP); // argument pointer
+    uint64_t upcallAddress = *(reinterpret_cast<uint64_t *>(reinterpret_cast<uint64_t>(metaData->upCallCommonDispatcher) + 8));
```

Comment:
Thanks for reminding me about this one. Julian had mentioned this in an earlier comment as well but I forgot to push this change.

We can store the dispatch routine's address in the function descriptor directly and then load it using `LAY` easily.

I've addressed this and your header file change in 0d9525b4e2a.

Reply:
@dchopra001 : I was actually questioning whether we should be loading a proper R5: env ptr prior to calling upcallAddress.

i.e. would it be more appropriate to have:

```C
    uint64_t upcallAddress = reinterpret_cast<uint64_t>(metaData->upCallCommonDispatcher);
    metaData->functionPtr[0] = upcallAddress;  # Store the full function descriptor for upCallCommonDispatcher
```

and instead of the LAY you added, we do:

```
LMG R5, R6, 0(R5)
BASR R7, R6
```

Reply:
Ah I see what you mean. I've fixed this in https://github.com/eclipse-openj9/openj9/commit/4d91569a48316cb8237a62b4037397d7761b079e.